src/BNO055.cpp

```diff
@@ -13,80 +13,69 @@ IMU::BNO055::BNO055(IO::I2C& i2C, uint8_t i2cSlaveAddress) : i2c(i2C) {
+        log::LOGGER.log(log::Logger::LogLevel::INFO, "Failed to detect IMU device with i2c and will quit initialization\r\n");
+    }
     log::LOGGER.log(log::Logger::LogLevel::INFO, "Device should be booted now... Checking if we can read...\r\n");
-
-    // Read the ID again, this is to make sure we are still connected to the correct device.
-    id = 0;
-    i2c.write(i2cAddress, 0x00);
     i2c.read(i2cAddress, &id);
-
     log::LOGGER.log(log::Logger::LogLevel::INFO, "ID Read 0x%x\r\n", id);
     if (id != BNO055_ID) {
         log::LOGGER.log(log::Logger::LogLevel::ERROR, "Failed first initialization... Trying again.\r\n");
 
         time::wait(1000);// Hold on for boot
-        i2c.write(i2cAddress, 0x00);
+        i2c.write(i2cAddress, 0x00); // keep reading id;
         i2c.read(i2cAddress, &id);
 
         if (id != BNO055_ID) {
-            log::LOGGER.log(log::Logger::LogLevel::ERROR, "Failed to initialize the IMU. Quitting initialization\r\n");
+            log::LOGGER.log(log::Logger::LogLevel::ERROR, "Failed to initialize the IMU. Quitting initialization.\r\n");
             return false;
         }
     }
 
     log::LOGGER.log(log::Logger::LogLevel::INFO, "Connected to i2c!\r\n");
 
-    // We need to wait another 50ms for the devie to figure itself out.
+    // We need to wait another 50ms for the device to figure itself out.
     time::wait(50);
 
-    log::LOGGER.log(log::Logger::LogLevel::INFO, "Setting power mode...\r\n");
-    // Set the power mode for the device, by writing to the Power mode address, followed by the normal power mode value.
-    uint8_t powerModeBytes[2] = {BNO055_PWR_MODE_ADDR, POWER_MODE_NORMAL};
-    i2c.write(i2cAddress, powerModeBytes, 2);
-    time::wait(10);
-
-    log::LOGGER.log(log::Logger::LogLevel::INFO, "Power mode set.\r\n");
-
-    // Reset the device page to 0x00.
-    log::LOGGER.log(log::Logger::LogLevel::INFO, "Reset device page.\r\n");
-    uint8_t pageBytes[2] = {BNO055_PAGE_ID_ADDR, 0x00};
-    i2c.write(i2cAddress, pageBytes, 2);
-
-    // Reset the system reset bits to 0x00, they were 0x20 because that's what we set them too earlier in the function.
-    log::LOGGER.log(log::Logger::LogLevel::INFO, "Reset the system reset.\r\n");
-    uint8_t systemResetResetBytes[2] = {BNO055_SYS_TRIGGER_ADDR, 0x00};
-    i2c.write(i2cAddress, systemResetResetBytes, 2);
-
-    time::wait(10);
+    uint8_t result;
+    // We read the ST_RESULT register that the startup POST self-test where the result for each sensor is put.
+    i2c.write(i2cAddress, BNO055_ST_RESULT);
+    i2c.read(i2cAddress, &result);
+    // All four LSB bits of result should be 1 for successful test
+    if ((result & 0x15) == 0x15){
+        log::LOGGER.log(log::Logger::LogLevel::ERROR, "Self-test failed. Quitting initialization.\r\n");
+        return false;
+    }else{
+        log::LOGGER.log(log::Logger::LogLevel::INFO, "Self-test passed, all sensors and microcontroller are functioning.\r\n");
+    }
```

Comment:
A self test is triggered every time the device boots up. It checks if all the sensors and microcontroller of the device is functioning correctly. By checking the results, we can avoid troubleshooting errors.